add fp16 support for mkldnn conv and deconv on CPU

CaoE opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99498
* #99497
* __->__ #99496

The PR is part of https://github.com/pytorch/pytorch/issues/97068, which is to add fp16 support for mkldnn conv and mkldnn deconv to leverage  avx_ne_convert, avx512-fp16, and amx-fp16 via the oneDNN library.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

CaoE committed:
add fp16 support for mkldnn conv on CPU

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/Utils.h:
```diff
@@ -74,6 +74,20 @@ inline bool mkldnn_bf16_device_check() {
+        "mkldnn_convolution_transpose: bf16 path needs the cpu support avx512bw, avx512vl and avx512dq");
+  } else if (input_t  == ScalarType::Half) {
+    TORCH_CHECK(mkldnn_fp16_device_check(),
+        "mkldnn_convolution_transpose: fp16 path needs the cpu support avx512fp16");
```
avx512fp16 doesn't have dot-product support. Are we sure that onednn has efficient conv/gemm support based on that?

mingfeima replied:
we need some benchmark number here.

CaoE replied:
> avx512fp16 doesn't have dot-product support. Are we sure that onednn has efficient conv/gemm support based on that?

Sorry for slow reply. I'm occupied by a urgent task. The performance should be worse than fp32 for computation bound ops on avx512fp16 platforms since there is internal fp16<->fp32 overhead.  Avx512_fp16 has compute instructions (fma, etc) , but those accumulate to f16, not f32. So unfortunately oneDNN cant use avx512_f16 compute instructions, only convert instructions are used.

jgong5 replied:
So, if this is only for the reference implementation, not necessarily we would rely on avx512fp16, right? Also, with avx512fp16, there might be accuracy concern since the accumulation is with f16. We shouldn't turn it on by default, right?

CaoE replied:
I think we would rely on avx512_fp16 since OneDNN requires the cpu has avx512_fp16 at least. Onednn does not use avx512_f16 to compute.  Only convert instructions are used to prevent accuracy issue.

jgong5 replied:
But why has it to depend on avx512_fp16 if only conversion instructions are used for ref implementation? Can f16c be used instead?

CaoE replied:
Currently oneDNN has no interest in supporting fp16 for anything earlier platforms than SPR so it requires the cpu has avx512_fp16 at least. oneDNN fp16 implements have much better performance compared to naive fp16 implements of PyTorch (shapes collect from RN50, not adequate testing) https://github.com/pytorch/pytorch/pull/99497#issue-1674040183 .  It benefits to call oneDNN with avx512_fp16.  We can add some conditions to avoid some slow cases of oneDNN, and ask onednn to improve the performance for such cases.

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"


The PR is part of https://github.com/pytorch/pytorch/issues/97068, which is to add fp16 support for mkldnn conv and mkldnn deconv to leverage avx512-fp16 or amx-fp16 via the oneDNN library.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
@cpuhrsch Could you please review this PR ? Thanks.

cpuhrsch commented on test/test_mkldnn.py:
```diff
@@ -299,27 +298,34 @@ def _test_conv_bf16_base(self, dim):
+        self._test_conv_lower_precision_base(dim=2, dtype=torch.bfloat16)
+        self._test_conv_lower_precision_base(dim=2, dtype=torch.half)
+
+    def test_conv3d_lower_precision(self):
```
nit: If you make these dtypes arguments to the tests and use a dtype decorator it'll be easier to see if one of them fails. Otherwise you'll get one failure for either one of the dtypes.

CaoE replied:
Thanks for the comments. I will improve such test cases in https://github.com/pytorch/pytorch/pull/99497 since it also involves these tests.

cpuhrsch reviewed (approved):
Generally looks good, just see one nit around using dtype + decorator to create separate, individual tests per dtype.

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"


The PR is part of https://github.com/pytorch/pytorch/issues/97068, which is to add fp16 support for mkldnn conv and mkldnn deconv to leverage  avx_ne_convert, avx512-fp16, and amx-fp16 via the oneDNN library.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
@pytorchbot merge

CaoE committed:
Update on "add fp16 support for mkldnn conv and deconv on CPU"


The PR is part of https://github.com/pytorch/pytorch/issues/97068, which is to add fp16 support for mkldnn conv and mkldnn deconv to leverage  avx_ne_convert, avx512-fp16, and amx-fp16 via the oneDNN library.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

CaoE commented:
@pytorchbot merge
